### DISCUSSION TITLE
Async in scheduled jobs

### DISCUSSION THREAD

**Discussion by avazon** [oqtane/oqtane.framework#3157] [Q&A] [answered]
Can a scheduled job be async?
From the sample developer module created "ExecuteJob" is overridden and is not async. There is a "ExecuteAsync" in HostedService base but if am correct this is more internal than meant to be used from modules.
I have a need for using HttpClient in a job and normally use async functions. The sample project/job for collecting Github statistics is similar but there it uses Getawaiter and is not async.
Anyone else has thoughts or experience around this?

**Comment by sbwalker** [comment, marked as answer]
@avazon The HostedServiceBase would need to include support for an async method (note the change from abstract to virtual methods which is a non-breaking change):

ie.

        // one of the following methods must be overridden
        public virtual string ExecuteJob(IServiceProvider provider)
        {
            return "";
        }

        public virtual Task<string> ExecuteJobAsync(IServiceProvider provider)
        {
            return Task.FromResult(string.Empty);
        }

and in the ExecuteAsync() method:

        notes += ExecuteJob(scope.ServiceProvider);
        notes += await ExecuteJobAsync(scope.ServiceProvider);

I think async support is a good idea for scheduled jobs

**Comment by sbwalker** [comment]
@avazon async support included in #3160 (which will be in next release)

**Reply by avazon** [reply]
Thank you
